CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs

```diff
@@ -150,5 +163,44 @@ public override void OnPhotoModeDeactivated()
             base.OnPhotoModeDeactivated();
             _weaponCrosshair.OnPhotoModeDeactivated();
         }
+
+        private bool IsPlayerAimingWithSniperRifle()
+        {
+            return Mission.CameraIsFirstPerson &&
+                   Input.IsKeyDown(InputKey.LeftShift) &&
+                   Input.IsKeyDown(InputKey.LeftMouseButton) &&
+                   Agent.Main.WieldedWeapon.Item.StringId.Contains("longrifle");
+        }
+
+        protected override void OnAgentControllerChanged(Agent agent, Agent.ControllerType oldController)
+        {
+            if (agent.Controller != Agent.ControllerType.Player || Agent.Main == null)
+            {
+                return;
+            }
+
+            _weaponCrosshair = new Crosshair(Mission, MissionScreen);
+            _weaponCrosshair.InitializeCrosshair();
+            _sniperScope = new SniperScope();
+            _isActive = true;
+
+            if (Agent.Main.IsAbilityUser() && (_abilityComponent = Agent.Main.GetComponent<AbilityComponent>()) != null)
+            {
+                _missionLogic = Mission.Current.GetMissionBehavior<AbilityManagerMissionLogic>();
+                _abilityComponent.CurrentAbilityChanged += (crosshair) =>
```

Comment:
You should remove the event listener in the Finalize method.

Reply:
yeah, you're totally right